### ISSUE TITLE
Validity of 'override' kwarg of MultiMethod.add(...) method ignored when adding first function 

### ISSUE THREAD

**Issue by m-charlton** [sunpy/sunpy#1991]
This is not a big problem, might want to clear up though. If this is the 
case then I can submit a fix along with some updated unit tests.
 
#### Description 

When I add a function to a `MultiMethod` instance, with *no* entries, and an invalid
value for the `override` keyword argument then the invalidity of this argument is
ignored and the function is successfully registered.

Adding a second function with an illegal value for `override` results in a 
`ValueError` as expected. Example:

```python

def eggs(): pass

def spam(): pass

mm = Multimethod(lambda *a: a)

# Does *not* raise ValueError('Invalid value for override'), it should
mm.add(spam, tuple(), override=56)

# Raises ValueError(...) as expected.
mm.add(eggs, tuple(), override=56)
```

It's a simple fix:

```python
# In sunpy/util/multimethod.py

def add(self, fun, types, override=SILENT)
    # ...
    # ...
    if override not in (SILENT, WARN, FAIL):
        raise ValueError("Invalid value '{0}' for override".format(override))

    # ...
    if override != SILENT:
        # ...
```

Not too keen about the duplication of the values for `override` but, can't think of a way around this.

**Comment by Cadair** [comment]
ping @segfaulthunter

**Comment by m-charlton** [comment]
Pinged florian.mayer@bitsrc.org

**Comment by segfaulthunter** [comment]
Looks like a small bug to me, yes.

**Comment by Cadair** [comment]
please send the patch @m-charlton another good spot. It's almost like we have tests for a reason XD

**Comment by m-charlton** [comment]
No worries. Should have the patch and updated tests ready for today or tomorrow,
at the very latest.

As an aside looking at some of the unit tests, I think that it may be a good 
idea to start mocking some of them. This would involve installing [mock](https://pypi.python.org/pypi/mock),
for python < 3.3 and [pytest-mock](https://pypi.python.org/pypi/pytest-mock)

I'd be interested in yours and the other developers opinions regarding this.

**Comment by Cadair** [comment]
@m-charlton some of the tests are going to be much harder to mock than others, but if you are up for it I am!

**Comment by m-charlton** [comment]
Agreed, sometimes mocking can result in tightly coupling test and application
code, which we'd obviously want to avoid, in order to achieve the holy grail of
100% code coverage. 

I'll start to investigate this in an isolated branch and report back when I've
something substantial to submit.
